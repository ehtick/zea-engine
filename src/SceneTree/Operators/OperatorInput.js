/** Class representing an operator input.
 */
class OperatorInput {
  /**
   * Create an operator input.
   * @param {string} name - The name value.
   */
  constructor(name) {
    this.__name = name
    this._param = undefined
    this._paramValueChanged = this._paramValueChanged.bind(this)
  }

  /**
   * The getName method.
   * @return {any} - The return value.
   */
  getName() {
    return this.__name
  }

  /**
   * Sets operator that owns this input. Called by the operator when adding inputs
   * @param {Operator} op - The operator object.
   */
  setOperator(op) {
    this._op = op
  }

  /**
   * Returns operator that owns this input.
   * @return {Operator} - The operator object.
   */
  getOperator() {
    return this._op
  }

  /**
   * Returns true if this input is connected to a parameter.
   * @return {boolean} - The return value.
   */
  isConnected() {
    return this._param != undefined
  }

  /**
   * The getParam method.
   * @return {any} - The return value.
   */
  getParam() {
    return this._param
  }

  /**
   * @private
   * The handler function for when the input paramter changes.
   * @param {object} event - The event object.
   */
  _paramValueChanged(event) {
    if (this._op) this._op.setDirty(this.__name)
  }

  /**
   * Assigns the Paramter to be used to provide the input value.
   * @param {Parameter} param - The param value.
   */
  setParam(param) {
    if (this._param) {
      this._param.off('valueChanged', this._paramValueChanged)
    }
    this._param = param
    if (this._param) {
      this._param.on('valueChanged', this._paramValueChanged)
    }
  }

  /**
   * The getValue method.
   * @return {any} - The return value.
   */
  getValue() {
    if (this._param) return this._param.getValue()
  }

  /**
   * The setValue method.
   * @param {any} value - The value param.
   */
  setValue(value) {
    if (this._param) {
      this._param.setValue(value)
    }
  }

  // ////////////////////////////////////////
  // Persistence

  /**
   * The toJSON method encodes this type as a json object for persistence.
   * @param {object} context - The context value.
   * @param {number} flags - The flags value.
   * @return {object} - Returns the json object.
   */
  toJSON(context, flags) {
    const paramPath = this._param ? this._param.getPath() : ''
    return {
      paramPath: context && context.makeRelative ? context.makeRelative(paramPath) : paramPath,
    }
  }

  /**
   * The fromJSON method decodes a json object for this type.
   * @param {object} j - The json object this item must decode.
   * @param {object} context - The context value.
   * @param {number} flags - The flags value.
   */
  fromJSON(j, context, flags) {
    if (j.paramPath) {
      // Note: the tree should have fully loaded by the time we are loading operators
      // even new items and groups should have been created. Operators and state machines
      // are loaded last.
      context.resolvePath(
        j.paramPath,
        (param) => {
          this.setParam(param)
        },
        (reason) => {
          console.warn("OperatorInput: '" + this.getName() + "'. Unable to connect to:" + j.paramPath)
        }
      )
    }
  }

  /**
   * The detach method is called when an operator is being removed from the scene tree.
   * It removes all connections to parameters in the scene.
   */
  detach() {
    // This function is called when we want to suspend an operator
    // from functioning because it is deleted and on the undo stack.
    // Once operators have persistent connections,
    // we will simply uninstall the output from the parameter.
    if (this._param) {
      this._param.off('valueChanged', this._paramValueChanged)
    }
  }

  /**
   * The reattach method can be called when re-instating an operator in the scene.
   */
  reattach() {
    this.detached = false
    if (this._param) {
      this._param.on('valueChanged', this._paramValueChanged)
    }
  }
}

export { OperatorInput }
