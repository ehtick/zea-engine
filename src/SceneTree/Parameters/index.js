export * from './Command.js'
export * from './Parameter.js'
export * from './MultiChoiceParameter.js'
export * from './BooleanParameter.js'
export * from './NumberParameter.js'
export * from './Vec2Parameter.js'
export * from './Vec3Parameter.js'
export * from './Vec4Parameter.js'
export * from './ColorParameter.js'
export * from './Mat3Parameter.js'
export * from './Mat4Parameter.js'
export * from './XfoParameter.js'
export * from './ImageParameter.js'
export * from './StringParameter.js'
export * from './CodeParameter.js'
export * from './FilePathParameter.js'
export * from './ListParameter.js'
export * from './StructParameter.js'
export * from './TreeItemParameter.js'
export * from './ItemSetParameter.js'
// export * from './KinematicGroupParameter.js';

export * from './MaterialFloatParam.js'
export * from './MaterialColorParam.js'
export * from './ProxyParameter.js'

// Note: can't be imported, because ParameterOwner imports these Parameters
// MaterialParameter imports MaterialLibraryManager which leads to circular imports
// export * from './MaterialParameter.js';
export * from './GeometryParameter.js'
