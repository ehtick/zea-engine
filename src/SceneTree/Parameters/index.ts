export * from './Parameter'
export * from './MultiChoiceParameter'
export * from './BooleanParameter'
export * from './NumberParameter'
export * from './Vec2Parameter'
export * from './Vec3Parameter'
export * from './Vec4Parameter'
export * from './ColorParameter'
export * from './QuatParameter'
export * from './Mat3Parameter'
export * from './Mat4Parameter'
export * from './XfoParameter'
export * from './ImageParameter'
export * from './StringParameter'
export * from './ListParameter'
// export * from './CodeParameter'
export * from './StructParameter'
export * from './TreeItemParameter'
export * from './ItemSetParameter'
export * from './OperatorOutputMode'

// Note: can't be imported, because ParameterOwner imports these Parameters
// export * from './MaterialParameter';
// export * from './MaterialFloatParam'
// export * from './MaterialColorParam'
export * from './GeometryParameter'
