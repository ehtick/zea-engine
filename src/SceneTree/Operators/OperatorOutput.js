import { ValueSetMode, ValueGetMode } from '../Parameters/index'

const OperatorOutputMode = {
  OP_WRITE: 0,
  OP_READ_WRITE: 1,
}


/** Class representing an operator output.
 */
class OperatorOutput {
  /**
   * Create an operator output.
   * @param {string} name - The name value.
   * @param {OperatorOutputMode} mode - The filterFn value.
   */
  constructor(name, mode) {
    this.__name = name
    this.__mode = mode
    this._param = undefined
    this.detached = false
  }

  /**
   * The getName method.
   * @return {any} - The return value.
   */
  getName() {
    return this.__name
  }

  setOperator(op) {
    this.__op = op
  }

  getOperator() {
    return this.__op
  }

  /**
   * The isConnected method.
   * @return {boolean} - The return value.
   */
  isConnected() {
    return this._param != undefined
  }

  /**
   * The getParam method.
   * @return {any} - The return value.
   */
  getParam() {
    return this._param
  }

  /**
   * The setParam method.
   * @param {any} param - The param value.
   */
  setParam(param) {
    this._param = param
    this._param.bindOperatorOutput(this)
  }

  /**
   * The getValue method.
   * @param {boolean} mode - The mode param.
   * @return {any} - The return value.
   */
  getValue(mode = ValueGetMode.OPERATOR_GETVALUE) {
    if (this._param) return this._param.getValue(mode)
  }

  /**
   * The setValue method.
   * Note: Sometimes outputs are used in places like statemachines,
   * where we would want the change to cause an event.
   * @param {any} value - The value param.
   * @param {boolean} mode - The mode value.
   */
  setValue(value, mode = ValueSetMode.OPERATOR_SETVALUE) {
    if (this._param) {
      this._param.setValue(value, mode)
    }
  }

  /**
   * The setClean method.
   * @param {any} value - The value param.
   */
  setClean(value) {
    if (this._param) {
      this._param.setClean(value)
    }
  }

  /**
   * The setDirty method.
   * @param {any} fn - The fn value.
   */
  setDirty() {
    if (this._param) {
      this._param.setDirty(this)
    }
  }

  setDirty() {
    if (this._param) {
      this._param.setDirty(this)
    }
  }

  /**
   * The setCleanFromOp method.
   * @param {any} value - The value param.
   */
  setCleanFromOp(value, op) {
    if (this._param) {
      this._param.setClean(value, op)
    }
  }

  /**
   * The removeCleanerFn method.
   * @param {any} fn - The fn value.
   */
  removeCleanerFn(fn) {
    if (this._param) this._param.removeCleanerFn(fn)
  }

  // ////////////////////////////////////////
  // Persistence

  /**
   * The toJSON method encodes this type as a json object for persistences.
   * @param {object} context - The context value.
   * @param {number} flags - The flags value.
   * @return {object} - Returns the json object.
   */
  toJSON(context, flags) {
    const paramPath = this._param ? this._param.getPath() : ''
    return {
      paramPath:
        context && context.makeRelative
          ? context.makeRelative(paramPath)
          : paramPath,
    }
  }

  /**
   * The fromJSON method decodes a json object for this type.
   * @param {object} j - The json object this item must decode.
   * @param {object} context - The context value.
   * @param {number} flags - The flags value.
   */
  fromJSON(j, context, flags) {
    if (j.paramPath) {
      // Note: the tree should have fully loaded by the time we are loading operators
      // even new items and groups should have been created. Operators and state machines
      // are loaded last.
      context.resolvePath(
        j.paramPath,
        param => {
          this.setParam(param)
        },
        reason => {
          console.warn(
            "Operator Output: '" +
              this.getName() +
              "'. Unable to load item:" +
              j.paramPath
          )
        }
      )
    }
  }

  // /**
  //  * The detach method.
  //  */
  // detach() {
  //   // This function is called when we want to suspend an operator
  //   // from functioning because it is deleted and on the undo stack.
  //   // Once operators have persistent connections,
  //   // we will simply uninstall the output from the parameter.
  //   this.detached = true
  // }

  // /**
  //  * The reattach method.
  //  */
  // reattach() {
  //   this.detached = false
  // }
}

/** Class representing an Xfo operator output.
 * @extends OperatorOutput
class XfoOperatorOutput extends OperatorOutput {
  /**
   * Create an Xfo operator output.
   * @param {string} name - The name value.
   * /
  constructor(name) {
    super(name)
  }

  /**
   * The getInitialValue method.
   * @return {any} - The return value.
   * /
  getInitialValue() {
    return this._initialParamValue
  }

  /**
   * The setParam method.
   * @param {any} param - The param value.
   * /
  setParam(param) {
    // Note: sometimes the param value is changed after binding.
    // e.g. The group Xfo is updated after the operator
    // that binds to it is loaded. It could also change if a user
    // Is adding items to the group using the UI. Therefore, the
    // initial Xfo needs to be updated.
    const init = () => {
      this._initialParamValue = param.getValue()
      if (this._initialParamValue.clone)
        this._initialParamValue = this._initialParamValue.clone()

      if (this._initialParamValue == undefined) throw new Error('WTF?')
    }
    init()
    // param.on('valueChanged', event => {
    //   if (
    //     event.mode == ValueSetMode.USER_SETVALUE ||
    //     event.mode == ValueSetMode.REMOTEUSER_SETVALUE ||
    //     event.mode == ValueSetMode.DATA_LOAD
    //   ) {
    //     init()
    //   }
    // })

    this._param = param
    this.emit('paramSet', { param })
  }
}
*/
export { OperatorOutput, OperatorOutputMode }
