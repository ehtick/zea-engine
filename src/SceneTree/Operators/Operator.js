import Registry from '../../Registry'
import { BaseItem } from '../BaseItem.js'
import { OperatorInput } from './OperatorInput'
import { OperatorOutput } from './OperatorOutput'

/**
 * Class representing an operator.
 *
 * @extends BaseItem
 */
class Operator extends BaseItem {
  /**
   * Create an operator.
   * @param {string} name - The name value.
   */
  constructor(name) {
    super(name)
    this.__inputs = new Map()
    this.__outputs = new Map()
  }

  /**
   * This method sets the state of the operator to dirty which propagates
   * to the outputs of this operator, and which may then propagate to other
   * operators. When the scene is cleaned, which usually is caused by rendering
   * then the chain of operators are cleaned by triggering evaluation.
   * @private
   */
  setDirty() {
    this.__outputs.forEach((output) => output.setDirty())
  }

  /**
   * This method can be overridden in derived classes
   * to perform general updates (see GLPass or BaseItem).
   *
   * @param {object} event
   * @private
   */
  __parameterValueChanged(event) {
    super.__parameterValueChanged(event)
    this.setDirty()
  }

  /**
   * The addInput method.
   * @param {string|OperatorInput} input - The name of the input, or the input object
   * @return {array} - The return value.
   */
  addInput(input) {
    if (typeof input == 'string') input = new OperatorInput(input)
    input.setOperator(this)
    this.__inputs.set(input.getName(), input)
    return input
  }

  /**
   * The removeInput method.
   * @param {string|OperatorInput} input - The name of the input, or the input object
   */
  removeInput(input) {
    if (typeof input == 'string') input = this.getInput(input)
    if (!(input instanceof OperatorInput)) {
      throw new Error('Invalid parameter for removeInput:', input)
    }
    if (input.getParam()) input.setParam(null)
    this.__inputs.delete(input.getName())
  }

  /**
   * Getter for the number of inputs in this operator.
   * @return {number} - Returns the number of inputs.
   */
  getNumInputs() {
    return this.__inputs.size
  }

  /**
   * The getInputByIndex method.
   * @param {number} index - The index value.
   * @return {object} - The return value.
   */
  getInputByIndex(index) {
    return Array.from(this.__inputs.values())[index]
  }

  /**
   * The getInput method.
   * @param {string} name - The name value.
   * @return {OperatorInput} - The return value.
   */
  getInput(name) {
    return this.__inputs.get(name)
  }

  /**
   * The addOutput method.
   * @param {string|OperatorOutput} output - The name of the output, or the output object
   * @return {array} - The return value.
   */
  addOutput(output) {
    if (typeof output == 'string') output = new OperatorOutput(output)
    output.setOperator(this)
    this.__outputs.set(output.getName(), output)
    return output
  }

  /**
   * The removeOutput method.
   * @param {string|OperatorOutput} output - The name of the output, or the output object
   */
  removeOutput(output) {
    if (typeof output == 'string') output = this.getOutput(output)
    if (!(output instanceof OperatorOutput)) {
      throw new Error('Invalid parameter for removeOutput:', output)
    }
    if (output.getParam()) output.setParam(null)
    this.__outputs.delete(output.getName())
  }

  /**
   * Getter for the number of outputs in this operator.
   * @return {number} - Returns the number of outputs.
   */
  getNumOutputs() {
    return this.__outputs.size
  }

  /**
   * The getOutputByIndex method.
   * @param {number} index - The index value.
   * @return {object} - The return value.
   */
  getOutputByIndex(index) {
    return Array.from(this.__outputs.values())[index]
  }

  /**
   * The getOutput method.
   * @param {string} name - The name value.
   * @return {OperatorOutput} - The return value.
   */
  getOutput(name) {
    return this.__outputs.get(name)
  }

  /**
   * The evaluate method.
   * Computes the values of each of the outputs based on the values of the inputs
   * and the values of outputs with mode OP_READ_WRITE.
   * This method must be implemented by all Operators.
   */
  evaluate() {
    throw new Error('Not yet implemented')
  }

  /**
   * The setValue method.
   * Note: FIXME Sometimes outputs are used in places like state-machines,
   * where we would want the change to cause an event.
   * Note: when a user sets a parameter value that is being driven by
   * an operator, the operator can propagate the value back up the chain
   * to its inputs.
   * @param {any} value - The value param.
   * @param {OperatorOutput} output - The output that we are receiving the setValue from
   * @return {number} - Returns the number of outputs.
   */
  setValue(value) {
    // TODO: Implement me for custom manipulations.
    return value
  }

  // ////////////////////////////////////////
  // Persistence

  /**
   * The toJSON method encodes this type as a json object for persistence.
   *
   * @param {object} context - The context value.
   * @return {object} - Returns the json object.
   */
  toJSON(context) {
    const j = super.toJSON(context)
    j.type = Registry.getBlueprintName(this)

    const inputs = []
    this.__inputs.forEach((input) => {
      inputs.push(input.toJSON(context))
    })
    j.inputs = inputs

    const outputs = []
    this.__outputs.forEach((output) => {
      outputs.push(output.toJSON(context))
    })
    j.outputs = outputs
    return j
  }

  /**
   * The fromJSON method decodes a json object for this type.
   *
   * @param {object} j - The json object this item must decode.
   * @param {object} context - The context value.
   */
  fromJSON(j, context) {
    super.fromJSON(j, context)

    if (j.inputs) {
      j.inputs.forEach((inputJson, index) => {
        let input
        if (inputJson.name) {
          input = this.getInput(inputJson.name)
          if (!input) {
            input = this.addInput(inputJson.name)
          }
        } else {
          input = this.getInputByIndex(index)
        }
        input.fromJSON(inputJson, context)
      })
    }
    if (j.outputs) {
      j.outputs.forEach((outputJson, index) => {
        let output
        if (outputJson.name) {
          output = this.getOutput(outputJson.name)
          if (!output) {
            output = this.addOutput(outputJson.name)
          }
        } else {
          output = this.getOutputByIndex(index)
        }
        output.fromJSON(outputJson, context)
      })
    }
  }

  /**
   * The detach method.
   */
  detach() {
    this.__inputs.forEach((input) => input.detach())
    this.__outputs.forEach((output) => output.detach())
  }

  /**
   * The reattach method.
   */
  reattach() {
    this.__inputs.forEach((input) => input.reattach())
    this.__outputs.forEach((output) => output.reattach())
  }

  /**
   * The rebind method.
   */
  rebind() {
    this.__outputs.forEach((output) => output.rebind())
  }
}

export { Operator }
