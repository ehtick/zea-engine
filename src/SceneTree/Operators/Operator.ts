/* eslint-disable @typescript-eslint/no-explicit-any */

import { BaseItem } from '../BaseItem'
import { OperatorInput } from './OperatorInput'
import { OperatorOutput } from './OperatorOutput'

/**
 * Class representing an operator.
 *
 * @extends BaseItem
 */
class Operator extends BaseItem {
  __inputs: Map<string, OperatorInput>
  __outputs: Map<string, OperatorOutput>
  /**
   * Create an operator.
   * @param name - The name value.
   */
  constructor(name: string = '') {
    super(name)
    this.__inputs = new Map()
    this.__outputs = new Map()
  }

  /**
   * This method sets the state of the operator to dirty which propagates
   * to the outputs of this operator, and which may then propagate to other
   * operators. When the scene is cleaned, which usually is caused by rendering
   * then the chain of operators are cleaned by triggering evaluation.
   * @private
   */
  setDirty(): void {
    this.__outputs.forEach((output: OperatorOutput) => output.setDirty())
  }

  /**
   * This method can be overridden in derived classes
   * to perform general updates (see GLPass or BaseItem).
   *
   * @param event
   * @private
   */
  protected parameterValueChanged(event: Record<string, unknown>): void {
    super.parameterValueChanged(event)
    this.setDirty()
  }

  /**
   * The addInput method.
   * @param input - The name of the input, or the input object
   * @return - The return value.
   */
  addInput(input: OperatorInput): OperatorInput {
    if (typeof input == 'string') input = new OperatorInput(input)
    else if (!(input instanceof OperatorInput)) {
      throw new Error(`addInput only accepts string or OperatorInput`)
    }
    input.setOperator(this)
    this.__inputs.set(input.getName(), input)
    this.setDirty()
    return input
  }

  /**
   * The removeInput method.
   * @param input - The name of the input, or the input object
   */
  removeInput(input: OperatorInput): void {
    if (typeof input == 'string') input = this.getInput(input)
    if (!(input instanceof OperatorInput)) {
      throw new Error(`removeInput only accepts string or OperatorInput`)
    }
    if (input.getParam()) input.setParam(undefined)
    this.__inputs.delete(input.getName())
  }

  /**
   * Getter for the number of inputs in this operator.
   * @return - Returns the number of inputs.
   */
  getNumInputs(): number {
    return this.__inputs.size
  }

  /**
   * The getInputByIndex method.
   * @param index - The index value.
   * @return - The return value.
   */
  getInputByIndex(index: number): Record<string, any> {
    return Array.from(this.__inputs.values())[index]
  }

  /**
   * The getInput method.
   * @param name - The name value.
   * @return - The return value.
   */
  getInput(name: string): OperatorInput {
    const input = this.__inputs.get(name)
    if (!input) throw `Couldn't find an Input with the name of '${name}'`
    return input
  }

  /**
   * The addOutput method.
   * @param output - The name of the output, or the output object
   * @return - The return value.
   */
  addOutput(output: OperatorOutput | string): OperatorOutput {
    if (typeof output == 'string') output = new OperatorOutput(output)
    else if (!(output instanceof OperatorOutput)) {
      throw new Error(`addOutput only accepts string or OperatorOutput`)
    }
    output.setOperator(this)
    // if (this.getOutput(output.getName())) throw new Error(`Operator output already exists ${output.getName()}`)
    this.__outputs.set(output.getName(), output)
    this.setDirty()
    return output
  }

  /**
   * The removeOutput method.
   * @param output - The name of the output, or the output object
   */
  removeOutput(output: OperatorOutput | string): void {
    if (typeof output == 'string') output = this.getOutput(output)
    if (!(output instanceof OperatorOutput)) {
      throw new Error(`removeOutput only accepts string or OperatorInput`)
    }
    if (output.getParam()) output.setParam()
    this.__outputs.delete(output.getName())
  }

  /**
   * Getter for the number of outputs in this operator.
   * @return - Returns the number of outputs.
   */
  getNumOutputs(): number {
    return this.__outputs.size
  }

  /**
   * The getOutputByIndex method.
   * @param index - The index value.
   * @return - The return value.
   */
  getOutputByIndex(index: number): OperatorOutput {
    return Array.from(this.__outputs.values())[index]
  }

  /**
   * The getOutput method.
   * @param name - The name value.
   * @return - The return value.
   */
  getOutput(name: string): OperatorOutput {
    const output = this.__outputs.get(name)
    if (!output) throw new Error(`Couldn't find an Output with the name of '${name}'`)
    return output
  }

  /**
   * The evaluate method.
   * Computes the values of each of the outputs based on the values of the inputs
   * and the values of outputs with mode OP_READ_WRITE.
   * This method must be implemented by all Operators.
   */
  evaluate(): void {
    throw new Error('Not yet implemented')
  }

  /**
   * When the value on a Parameter is modified by a user by calling 'setValue,
   * then if any operators are bound, the value of the Parameter cannot be modified
   * directly as it is the result of a computation. Instead, the Parameter calls
   * 'backPropagateValue' on the Operator to cause the Operator to handle propagating
   * the value to one or more of its inputs.
   * to its inputs.
   * @param value - The value param.
   * @return - The modified value.
   */
  backPropagateValue(value: unknown): unknown {
    // TODO: Implement me for custom manipulations.
    return value
  }

  // ////////////////////////////////////////
  // Persistence

  /**
   * The toJSON method encodes this type as a json object for persistence.
   *
   * @param context - The context value.
   * @return - Returns the json object.
   */
  toJSON(context?: Record<string, any>): Record<string, any> {
    const j = super.toJSON(context)
    ;(j as any).type = this.getClassName()
    const inputs: any = []
    this.__inputs.forEach((input: any) => {
      inputs.push(input.toJSON(context))
    })
    ;(j as any).inputs = inputs
    const outputs: any = []
    this.__outputs.forEach((output: any) => {
      outputs.push(output.toJSON(context))
    })
    ;(j as any).outputs = outputs
    return j
  }

  /**
   * The fromJSON method decodes a json object for this type.
   *
   * @param j - The json object this item must decode.
   * @param context - The context value.
   */
  fromJSON(j: Record<string, any>, context?: Record<string, any>): void {
    super.fromJSON(j, context)
    if (j.inputs) {
      j.inputs.forEach((inputJson: any, index: any) => {
        let input
        if (inputJson.name) {
          input = this.getInput(inputJson.name)
          if (!input) {
            input = this.addInput(inputJson.name)
          }
        } else {
          input = this.getInputByIndex(index)
        }
        ;(input as OperatorInput).fromJSON(inputJson, context)
      })
    }
    if (j.outputs) {
      j.outputs.forEach((outputJson: any, index: any) => {
        let output
        if (outputJson.name) {
          output = this.getOutput(outputJson.name)
          if (!output) {
            output = this.addOutput(outputJson.name)
          }
        } else {
          output = this.getOutputByIndex(index)
        }
        ;(output as OperatorOutput).fromJSON(outputJson, context)
      })
    }
  }

  /**
   * The detach method.
   */
  detach(): void {
    this.__inputs.forEach((input: OperatorInput) => input.detach())
    this.__outputs.forEach((output: OperatorOutput) => output.detach())
  }

  /**
   * The reattach method.
   */
  reattach(): void {
    this.__inputs.forEach((input: OperatorInput) => input.reattach())
    this.__outputs.forEach((output: OperatorOutput) => output.reattach())
  }

  /**
   * The rebind method.
   */
  rebind(): void {
    this.__outputs.forEach((output: OperatorOutput) => output.rebind())
  }
}

export { Operator }
