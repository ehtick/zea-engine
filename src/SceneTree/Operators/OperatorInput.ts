/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { Parameter } from '@SceneTree/Parameters'
import { EventEmitter } from '@Utilities/EventEmitter'
import { Operator } from './Operator'

/** Class representing an operator input.
 * @extends EventEmitter
 */
class OperatorInput extends EventEmitter {
  __name: string
  _op: Operator
  _param?: Parameter
  detached: boolean

  /**
   * Create an operator input.
   * @param {string} name - The name value.
   */
  constructor(name: string) {
    super()
    this.__name = name
    this.paramValueChanged = this.paramValueChanged.bind(this)
  }

  /**
   * The getName method.
   * @return {string} - The return value.
   */
  getName(): string {
    return this.__name
  }

  /**
   * Sets operator that owns this input. Called by the operator when adding inputs
   * @param {Operator} op - The operator object.
   */
  setOperator(op: Operator): void {
    this._op = op
  }

  /**
   * Returns operator that owns this input.
   * @return {Operator} - The operator object.
   */
  getOperator(): Operator {
    return this._op
  }

  /**
   * Returns true if this input is connected to a parameter.
   * @return {boolean} - The return value.
   */
  isConnected(): boolean {
    return this._param != undefined
  }

  /**
   * The getParam method.
   * @return {Parameter} - The return value.
   */
  getParam(): Parameter | undefined {
    return this._param
  }

  /**
   * @private
   * The handler function for when the input paramter changes.
   * @param {object} event - The event object.
   */
  protected paramValueChanged = (event: Record<string, any>): void => {
    if (this._op) this._op.setDirty()
  }

  /**
   * Assigns the Paramter to be used to provide the input value.
   * @param {Parameter} param - The param value.
   */
  setParam(param?: Parameter): void {
    if (this._param) {
      this._param.off('valueChanged', this.paramValueChanged)
    }
    this._param = param
    if (this._param) {
      this._param.on('valueChanged', this.paramValueChanged)
    }
    this.emit('paramSet', { param: this._param })
  }

  /**
   * The getValue method.
   * @return {unknown} - The return value.
   */
  getValue(): unknown {
    if (this._param) return this._param.getValue()
  }

  /**
   * The setValue method.
   * @param {unknown} value - The value param.
   */
  setValue(value: unknown): void {
    if (this._param) {
      this._param.setValue(value)
    }
  }

  // ////////////////////////////////////////
  // Persistence

  /**
   * The toJSON method encodes this type as a json object for persistence.
   *
   * @param {Record<string, any>} context - The context value.
   * @return {{ name: string; paramPath: any }} - Returns the json object.
   */
  toJSON(context?: Record<string, any>): { name: string; paramPath: any } {
    const paramPath = this._param ? this._param.getPath() : ''
    return {
      name: this.__name,
      paramPath: context && context.makeRelative ? context.makeRelative(paramPath) : paramPath,
    }
  }

  /**
   * The fromJSON method decodes a json object for this type.
   * @param {Record<string, any>} j - The json object this item must decode.
   * @param {Record<string, any>} context - The context value.
   */
  fromJSON(j: Record<string, any>, context?: Record<string, any>): void {
    if (j.paramPath) {
      // Note: the tree should have fully loaded by the time we are loading operators
      // even new items and groups should have been created. Operators and state machines
      // are loaded last.
      context?.resolvePath(
        j.paramPath,
        (param: any) => {
          this.setParam(param)
        },
        (reason: any) => {
          console.warn("OperatorInput: '" + this.getName() + "'. Unable to connect to:" + j.paramPath)
        },
      )
    }
  }

  /**
   * The detach method is called when an operator is being removed from the scene tree.
   * It removes all connections to parameters in the scene.
   */
  detach(): void {
    // This function is called when we want to suspend an operator
    // from functioning because it is deleted and on the undo stack.
    // Once operators have persistent connections,
    // we will simply uninstall the output from the parameter.
    if (this._param) {
      this._param.off('valueChanged', this.paramValueChanged)
    }
  }

  /**
   * The reattach method can be called when re-instating an operator in the scene.
   */
  reattach(): void {
    this.detached = false
    if (this._param) {
      this._param.on('valueChanged', this.paramValueChanged)
    }
  }
}

export { OperatorInput }
