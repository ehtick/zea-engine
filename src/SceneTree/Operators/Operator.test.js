/* eslint-disable require-jsdoc */
import { Operator } from './Operator'
import { NumberParameter } from '../Parameters/NumberParameter'
import { OperatorInput } from './OperatorInput'
import { OperatorOutput, OperatorOutputMode } from './OperatorOutput'

class AddFloatsOperator extends Operator {
  constructor(name) {
    super(name)
    this.addInput(new OperatorInput('A'))
    this.addInput(new OperatorInput('B'))
    this.addOutput(new OperatorOutput('C'))
  }
  
  evaluate() {
    const a = this.getInput('A').getValue()
    const b = this.getInput('B').getValue()
    this.getOutput('C').setClean(a + b)
  }
}


// Note: this operatore modifies a value in the output attribute
// By reading and then changing. This feature allows us to combine operators
// to compute complex results. (See BoundingBox operators).
class ScaleFloatOperator extends Operator {
  constructor(name) {
    super(name)
    this.addInput(new OperatorInput("ScaleValue"));
    this.addOutput(new OperatorOutput("Result", OperatorOutputMode.OP_READ_WRITE));
  }

  evaluate() {
    let scaleValue = 1.0
    const inParam = this.getInput("ScaleValue").getParam()
    if (inParam) {
      scaleValue = inParam.getValue()
    }
    // Read the value, modify and return.
    const value = this.getOutput("Result").getValue()
    this.getOutput("Result").setClean(value * scaleValue)
  }
}

describe('Operator', () => {
  it('AddFloatsOperator', () => {
    const addOperator = new AddFloatsOperator()

    const aParam = new NumberParameter('A')
    const bParam = new NumberParameter('B')
    const myParam = new NumberParameter('MyParam')

    addOperator.getInput("A").setParam(aParam)
    addOperator.getInput("B").setParam(bParam)
    addOperator.getOutput("C").setParam(myParam)

    aParam.setValue(3)
    bParam.setValue(2.5)

    expect(myParam.isDirty()).toEqual(true)
    expect(myParam.getValue()).toEqual(5.5)
    expect(myParam.isDirty()).toEqual(false)
  })

  it('ScaleFloatOperator', () => {
    const scaleOperator = new ScaleFloatOperator()

    const scaleParam = new NumberParameter('A', 2)
    const resultParam = new NumberParameter('MyParam', 3)

    scaleOperator.getInput("ScaleValue").setParam(scaleParam)
    scaleOperator.getOutput("Result").setParam(resultParam)

    expect(resultParam.isDirty()).toEqual(true)
    expect(resultParam.getValue()).toEqual(6)
    expect(resultParam.isDirty()).toEqual(false)

    scaleParam.setValue(4)

    expect(resultParam.isDirty()).toEqual(true)
    expect(resultParam.getValue()).toEqual(24)
    expect(resultParam.isDirty()).toEqual(false)
  })

  it('AddScaleFloatOperator', () => {
    const addOperator = new AddFloatsOperator()
    const aParam = new NumberParameter('A', 2)
    const bParam = new NumberParameter('B', 3.5)
    const myParam = new NumberParameter('AddScaleFloatOperator-MyParam')

    addOperator.getInput("A").setParam(aParam)
    addOperator.getInput("B").setParam(bParam)
    addOperator.getOutput("C").setParam(myParam)

    const scaleOperator = new ScaleFloatOperator()
    const scaleParam = new NumberParameter('A', 2)

    scaleOperator.getInput("ScaleValue").setParam(scaleParam)
    scaleOperator.getOutput("Result").setParam(myParam)

    expect(myParam.isDirty()).toEqual(true)
    expect(myParam.getValue()).toEqual(11)
    expect(myParam.isDirty()).toEqual(false)

    scaleParam.setValue(1)

    expect(myParam.isDirty()).toEqual(true)
    expect(myParam.getValue()).toEqual(5.5)
    expect(myParam.isDirty()).toEqual(false)
  })
})
