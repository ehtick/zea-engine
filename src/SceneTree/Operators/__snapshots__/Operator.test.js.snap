// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Operator save to JSON (serialization). 1`] = `"{\\"name\\":\\"\\",\\"type\\":\\"AddFloatsOperator\\",\\"inputs\\":[{\\"name\\":\\"A\\",\\"paramPath\\":[\\"Foo\\",\\"A\\"]},{\\"name\\":\\"B\\",\\"paramPath\\":[\\"Foo\\",\\"B\\"]}],\\"outputs\\":[{\\"name\\":\\"C\\",\\"paramPath\\":[\\"Foo\\",\\"MyParam\\"],\\"paramBindIndex\\":0}]}"`;
